### ISSUE TITLE
Ignoring unsupported options: build in Ignoring unsupported options: build

### ISSUE THREAD

**Issue by akhildangore** [docker/docker#29940]
Hello All,

Docker version: v1.13.0-rc4

I am getting error : **Ignoring unsupported options: build**, Does it mean it will not create build from Dockerfile ?

And also getting same error for network_mode: **Ignoring unsupported options: network_mode**.

Below is command:
DOCKER_IMAGE=akhil123 docker stack deploy -c docker-compose.yml foo

Many thanks in advance.

**Comment by thaJeztah** [comment]
I replied to your comment here; https://github.com/docker/docker/issues/29133#issuecomment-270901505

> The `docker stack deploy` feature, _by design_ does not perform builds. Reason for this is that a _build_ is performed on the host that the command is run from, so the image will only be available on _that_ node. When deploying that image in a Swarm, the service cannot be started on other nodes. To deploy services, make sure the image you're deploying is pushed to a registry (or for testing; make sure the image is available on every node in the swarm)

**Comment by D3nn** [comment]
Outside of the "build" directive (which I agree doesn't make much sense here), I'd like to request that the "network_mode" from the docker-compose file be supported.  I know that there's already been some discussion around this from those who'd like to use it for ```network_mode:host``` capabilities, but personally I'm more interested in the ```network_mode:"service:<svc name>"``` and ```network_mode:"container:<container name|id>"``` capabilities.

I'd also like to see this supported in the ```docker service create``` command or at least have the --network option be able to include the same semantics as the network_mode does in the compose file.

The use cases I see for this are:
 * launching containers that share the same network stack so that applications can communicate with each other over a loopback interface (and don't have to worry about service discovery)
 * launching a container which contains a network toolset so that I can do some network diagnostics on processes that run in containers derived from "scratch" (such as containers running go programs).

The first use case would definitely benefit from reviving container affinity constraints from the old swarm, but that could be worked around for now using the supported placement constraints from v3 compose format.

**Comment by D3nn** [comment]
Should I open another issue for network_mode, or is the comment in this issue sufficient?  (Just checking process-wise)

**Comment by thaJeztah** [comment]
Yes, please open a separate issue. FWIW, `network_mode:"container:<container name|id>"` will be troublesome, as it basically says that _all_ tasks of a service (and the "dependent service") _must_ live on the same host

**Comment by trajano** [comment]
If `build` is not supported in `stack deploy` when is it used then?  Maybe https://docs.docker.com/compose/compose-file/#build should just be removed?

**Comment by cpuguy83** [comment]
@trajano It's used with `docker-compose`.

**Comment by wesleyvicthor** [comment]
@trajano maybe just a proper docs update with the provided infos from @thaJeztah

**Comment by yoplait** [comment]
Then the problem is that you need a file to deploy (docker stack deploy) and another file to build (docker-compose)
And the problem on this is that you need to maintain double quantity of files, as if I am using version: "3.2" to support in my deployments ports mode host, and I need to be on version: '3' and without that params for the builds in docker-compose …

**Comment by vdemeester** [comment]
> Then the problem is that you need a file to deploy (docker stack deploy) and another file to build (docker-compose)

That's not really true, as you can use the same file for both (i.e. `docker-compose build`, `docker-compose push` to build and push and `docker stack deploy` to deploy).

**Comment by yoplait** [comment]
Thanks @vdemeester you are correct I just realized that I was not in the latest docker-compose … I was thinking that after to write my comment … sorry!

**Comment by ntwrkguru** [comment]
Actually, this is a lame decision. How hard would it be to build the image on all nodes? This complicates my life greatly as we provide stacks as a deliverable to customers and now will have to build and push EVERY variant of every image instead of building it at deploy as we did with compose. Thankfully, I can still build via Ansible and hope that the `docker_stack` module is released and full-featured soon.

**Comment by chiefy** [comment]
Using UCP and running into the same situation as others, my use case is I have a pretty simple docker image that needs to be built on `docker stack deploy ...` and I really don't care if it needs built on all nodes. In this case it only needs built on manager nodes really since it's a proxy.

**Comment by francisliyy** [comment]
@yoplait , How you support mode host in your deployment, could you show your docker compose file?
Thank you .

**Comment by jnorthrup** [comment]
it looks like the decision to HARD FORK the docker-compose, and then make BREAKING CHANGES willy nilly is the docker strategy here.

why does docker compose EXPRESSLY CITE docker-compose syntax without its own wiki/docs/whatever and then force the consumer to start hunting for error text keywords like 

`Ignoring unsupported options: build`

who came up with this strategy? i dont see a single mention of "WHITELISTED DOCKER-COMPOSE DEPLOY KEYWORDS when i fumble through the official docker swarm migration steps
